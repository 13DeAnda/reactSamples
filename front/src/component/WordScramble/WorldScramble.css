
.WordScramble .letterBox{
  height: 40px;
  width: 40px;
  display: inline;
  padding: 18px;
  margin: 15px;
  border-radius: 5px;
  background-color: #ebdcf5b8;
  font-weight: bold;
}

.WordScramble .scrambledWordContainer{
  margin-top: 30px;
  margin-bottom: 30px;
}

.WordScramble .foundLetter{
  background: #610a9ab8;
  color: white;
}


.WordScramble .skipButton{
  font-size: 11px;
  padding: 5px;
  margin-right: 24px;
}

.WordScramble .score{
  font-size: 20px;
}

.WordScramble .userTextBox{
    border-radius: 5px;
    width: 300px;
    height: 50px;
    font-size: 20px;
    padding: 5px;
    text-align: center;
    border-color: purple;

    border-top: none;
    border-left: none;
    border-right: none;
}
.WordScramble .counterContainer{
  font-size: 20px;
}

.WordScramble .correct{
  color: #5e2184;
  font-size: 30px;
  text-shadow: 1px 1px #a99999;
}
.WordScramble .incorrect{
  color: red;
  font-size: 30px;
  text-shadow: 1px 1px #a99999;
}

.WordScramble input:focus{
  outline-color: white;
}

.modal-content{
  margin-top: 30% !important;
  font-size: 20px;

}