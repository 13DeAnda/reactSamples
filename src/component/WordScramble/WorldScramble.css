
.WordScramble{
  padding-top: 50px;
  padding-bottom: 50px;
}
.letterBox{
  height: 40px;
  width: 40px;
  display: inline;
  padding: 18px;
  margin: 15px;
  border-radius: 5px;
  background-color: beige;
}

.scrambledWordContainer{
  margin-top: 30px;
}
.text-center{
  text-align: center;
}

.foundLetter{
  background: green;
}