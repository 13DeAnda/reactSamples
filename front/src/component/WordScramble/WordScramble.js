import React from 'react';
import axios from 'axios';
import _ from 'lodash';
import '../../shared/shared.css';
import './WorldScramble.css';
import { Grid, Row, Modal} from 'react-bootstrap';
import ReactInterval from 'react-interval';
export default class WordScramble extends React.Component {

  constructor(props) {
    super(props);
    this.scrambleText = this.scrambleText.bind(this);
    this.onChange = this.onChange.bind(this);
    this.checkForLetter = this.checkForLetter.bind(this);
    this.checkIfCorrect = this.checkIfCorrect.bind(this);
    this.interval = this.interval.bind(this);
    this.getWord = this.getWord.bind(this);
    this.state = {
      userWord: "",
      scrambledWord: "",
      scrambledWordObj: [],
      scrambledWordObjCopy: [],
      usedLetters: [],
      correctWord: "",
      score: 0,
      timerCount: 0,
      enabled: false,
      message: ""
    };
  }

  componentDidMount(){
    this.getWord("responsible");
  }

  render() {
    var {enabled, scrambledWord, scrambledWordObj,
         message, usedLetters, userWord, timerCount, score} = this.state
    return (
      <Grid className="container WordScramble text-center">
        <Row className={scrambledWord? 'text-right' : 'hidden'}>
            <div className='counterContainer'>
              Time left: {timerCount}
              <ReactInterval timeout={1000} enabled={enabled} callback={this.interval} />
            </div>
            <button className='skipButton' onClick={this.getWord}> Skip Word </button>
        </Row>
        <Row className={scrambledWord? '' : 'hidden'}>
          <input type="text" className='userTextBox' placeholder='Type your word' value={userWord} onChange={this.onChange}/>
        </Row>
       <Row className="scrambledWordContainer">
          {usedLetters.map(function(item, index){
            return <div className ='letterBox foundLetter' key={index}>{item.letter} </div>
          })}

          {scrambledWordObj.map(function(item, index){
            return <div className = {item.used? 'hidden' : 'letterBox'} key={index}>{item.letter} </div>
          })}
        </Row>
        <Row className={score.length > 0 ? 'score' : 'hidden'}>
          matches:  {score}
        </Row>
        <Row className={message}>
          {message.length >0?
            message: null}
        </Row>

        <Modal show={this.state.gameOver} onHide={this.getWord}>
          <Modal.Body className='score text-center'>
              <div className="content"> Your Score: {this.state.score} </div> <br/>
              <button className='restartButton' onClick={this.getWord}>Start Over</button>
          </Modal.Body>
        </Modal>
      </Grid>
    );
  }
  //logic
  getWord(word) {
    //expects a word when its called on test only
    if(word && typeof word === "string"){
      this.scrambleText(word);
    }
    else{
      var that = this;
      axios.get('http://api.wordnik.com/v4/words.json/randomWord?hasDictionaryDef=true&includePartOfSpeech=adverb&excludePartOfSpeech=verb&minCorpusCount=0&maxCorpusCount=-1&minDictionaryCount=1&maxDictionaryCount=-1&minLength=5&maxLength=8&api_key=a2a73e7b926c924fad7001ca3111acd55af2ffabf50eb4ae5')
      .then(function(response){
        if(response.state === 200){
          that.scrambleText(response.data.word);
        }
      })
      .catch(function (error) {
        that.setState({message: "An error occurred with the API request",
                       correctWord: null,
                       scrambledWordObj: [],
                       scrambledWord: null});
      });
    }
  }

  scrambleText (word){
    var scrambledWord = "";
    var usedIndex = {};
    var scrambledWordTemp = [];

    while(scrambledWord.length !== word.length){
      var randomIndex = Math.floor(Math.random() * word.length) + 0;
      if(!usedIndex[randomIndex]){
        usedIndex[randomIndex] = true;
        scrambledWord += word[randomIndex];
        scrambledWordTemp.push({letter: word[randomIndex], used: false})
      }
    }

    var toUpdate = {scrambledWord: scrambledWord,
                   correctWord: word,
                   scrambledWordObjCopy: scrambledWordTemp,
                   scrambledWordObj: scrambledWordTemp,
                   gameOver: false};

    if(!this.state.enabled){
      toUpdate.enabled = true;
      toUpdate.timerCount = 60;
    }
    this.setState(toUpdate);
    return scrambledWord;
  }

  checkForLetter(newText){
    var inWordText = "";
    var usedLetters = [];
    var scrambledWordObjTemp = _.cloneDeep(this.state.scrambledWordObjCopy);
    _.forEach(newText, function(letter){

      _.forEach(scrambledWordObjTemp, function(item){
        if(!item.used && item.letter === letter){
          item.used = true;
          usedLetters.push({letter: item.letter});
          inWordText += item.letter;
          return false
        }
      });
    });
    this.setState({userWord: inWordText, usedLetters: usedLetters, scrambledWordObj: scrambledWordObjTemp});
    this.checkIfCorrect(newText);
  }

  checkIfCorrect(newText){
    if(newText === this.state.correctWord){
      var score = _.cloneDeep(this.state.score);
      this.setState({score: ++score, message: "correct"});
      setTimeout(function() {
        this.setState({message: "", usedLetters: [], userWord: ""});
        this.getWord();
      }.bind(this), 2000);
    }
    else if(newText.length === this.state.correctWord.length){
      this.setState({message: "incorrect"});
      setTimeout(function() {
        this.setState({message: ""});
      }.bind(this), 2000);
    }
  }

  interval(){
    if(this.state.timerCount > 0 ){
      this.setState({timerCount: this.state.timerCount - 1});
    }
    else{
      this.setState({enabled: false, gameOver: true});
    }
  }

  onChange(e){
    const newText = e.target.value;
    this.checkForLetter(newText);
  }
}
