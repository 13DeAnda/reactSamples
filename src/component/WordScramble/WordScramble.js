import React from 'react';
import axios from 'axios';
import _ from 'lodash';
import './WorldScramble.css';
import { Grid, Row } from 'react-bootstrap';
import ReactInterval from 'react-interval';

class WordScramble extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      userWord: "",
      scrambledWord: "",
      scrambledWordObj: [],
      scrambledWordObjCopy: [],
      usedLetters: [],
      correctWord: "",
      score: 0,
      timerCount: 60,
      enabled: false
    };
  }

  getWord = () =>  {
    var that = this;
    axios.get('http://api.wordnik.com/v4/words.json/randomWord?hasDictionaryDef=true&includePartOfSpeech=adverb&excludePartOfSpeech=verb&minCorpusCount=0&maxCorpusCount=-1&minDictionaryCount=1&maxDictionaryCount=-1&minLength=5&maxLength=8&api_key=a2a73e7b926c924fad7001ca3111acd55af2ffabf50eb4ae5')
      .then(function(response){
        console.log("the word", response.data.word);
        that.scrambleText(response.data.word);
      })
  }

  scrambleText = (word) =>{
    var scrambledWord = "";
    var usedIndex = {};
    var scrambledWordTemp = [];

    while(scrambledWord.length !== word.length){
      var randomIndex = Math.floor(Math.random() * word.length) + 0;
      if(!usedIndex[randomIndex]){
        usedIndex[randomIndex] = true;
        scrambledWord += word[randomIndex];
        scrambledWordTemp.push({letter: word[randomIndex], used: false})
      }
    }

    var toUpdate = {scrambledWord: scrambledWord, 
                   correctWord: word, 
                   scrambledWordObjCopy: scrambledWordTemp,
                   scrambledWordObj: scrambledWordTemp};

    if(!this.state.enabled){
      toUpdate.enabled = true;
    }
    this.setState(toUpdate);
  }

  onChange = (e) => {
    const newText = e.target.value;
    this.checkForLetter(newText);
  }

  checkForLetter = (newText) => {
    var inWordText = "";
    var usedLetters = [];
    var scrambledWordObjTemp = _.cloneDeep(this.state.scrambledWordObjCopy);
    _.forEach(newText, function(letter){

      _.forEach(scrambledWordObjTemp, function(item){
        if(!item.used && item.letter === letter){
          item.used = true;
          usedLetters.push({letter: item.letter});
          inWordText += item.letter;
          return false
        }
      });     
    });
    this.setState({userWord: inWordText, usedLetters: usedLetters, scrambledWordObj: scrambledWordObjTemp});
    this.checkIfCorrect(newText);
  }

  checkIfCorrect = (newText) =>{
    if(newText === this.state.correctWord){
      var score = _.cloneDeep(this.state.score);
      this.setState({score: ++score, message: "correct!"});
      setTimeout(function() { 
        this.setState({message: "", usedLetters: [], userWord: ""});
        this.getWord();
      }.bind(this), 1000);
    }
    else if(newText.length === this.state.correctWord.length){
      this.setState({message: "Incorrect"});
    }
  }

  interval = () =>{
    if(this.state.timerCount > 0 ){
      this.setState({timerCount: this.state.timerCount - 1});
    }
    else{
      this.setState({enabled: false});
    }
  }

  render() {
    return (
      <Grid className="container-fluid WordScramble">
        {this.state.scrambledWord.length === 0 && !this.state.enabled? this.getWord() : null}
        <Row className="text-rigth">
            <div className='counterContainer'>
              Time left: {this.state.timerCount}
              <ReactInterval timeout={1000} enabled={this.state.enabled} callback={this.interval} />
            </div>
            <button className='skipButton' onClick={this.getWord}> Skip Word </button> 
        </Row>
        <Row className="text-center"> 
          <input type="text" className='userTextBox' placeholder='Type your word' value={this.state.userWord} onChange={this.onChange}/>
        </Row>
       <Row className="scrambledWordContainer text-center">
          {this.state.usedLetters.map(function(item, index){
            return <div className ='letterBox foundLetter' key={index}>{item.letter} </div>
          })}

          {this.state.scrambledWordObj.map(function(item, index){
            return <div className = {item.used? 'hide' : 'letterBox'} key={index}>{item.letter} </div>
          })}
        </Row>
        <Row className="text-center score">
          matches:  {this.state.score}
        </Row>
      </Grid>
    );
  }
}
export default WordScramble;
