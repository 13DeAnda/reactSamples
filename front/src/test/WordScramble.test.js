import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import WordScramble from '../component/WordScramble/WordScramble';

const wrapper = shallow(<WordScramble />);
const instance = wrapper.instance();

const correctWord = "test";

describe('WordScramble component', () => {

  //unit test
  it('expects  scrambleText() result the same length', () => {
    var scrambledText = instance.scrambleText(correctWord);
    expect(scrambledText.length).toEqual(correctWord.length);
  });

  it('expects the scrambledText result not equal the original', () => {
    var scrambledText = instance.scrambleText(correctWord);
    expect(scrambledText).not.toEqual(correctWord);
  });

  it('expects the getWord() result to not be the same length', () => {
    var scrambledText = instance.getWord(correctWord);
    expect(scrambledText.length).toEqual(correctWord.length);
  });

  it('expects the getWord() to not be the same length', () => {
    var scrambledText = instance.getWord(correctWord);
    expect(scrambledText).not.toEqual(correctWord);
  });

  //UI testing
  it('should render snapshot', () => {
    const component = renderer.create(
      <WordScramble />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });



  it('types on the textbox and expects usedWords to increase and display on view', () => {
    var userText = "te";
    wrapper.find('.userTextBox').simulate('change', {target: {value: userText}});

    expect(wrapper.state().userWord).toEqual(userText);
  });

  it('should not display letters not in the scrambled  word', () => {
    var userText = "km";
    wrapper.find('.userTextBox').simulate('change', {target: {value: userText}});
    expect(wrapper.state().userWord.length).not.toEqual(userText.length);
  });


  it('should display changed score if correct word is typed', () => {
    wrapper.find('.userTextBox').simulate('change', {target: {value: correctWord}});
    expect(wrapper.state().score).toEqual(1);
  });

});
