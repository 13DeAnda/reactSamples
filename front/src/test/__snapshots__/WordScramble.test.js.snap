// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordScramble component should render snapshot 1`] = `
<div
  className="container WordScramble text-center container"
>
  <div
    className="text-right row"
  >
    <div
      className="counterContainer"
    >
      Time left: 
      60
    </div>
    <button
      className="skipButton"
      onClick={[Function]}
    >
       Skip Word 
    </button>
  </div>
  <div
    className="row"
  >
    <input
      className="userTextBox"
      onChange={[Function]}
      placeholder="Type your word"
      type="text"
      value=""
    />
  </div>
  <div
    className="scrambledWordContainer row"
  >
    <div
      className="letterBox toUseLetter"
      onClick={[Function]}
      value="e"
    >
      e
    </div>
    <div
      className="letterBox toUseLetter"
      onClick={[Function]}
      value="e"
    >
      e
    </div>
    <div
      className="letterBox toUseLetter"
      onClick={[Function]}
      value="r"
    >
      r
    </div>
    <div
      className="letterBox toUseLetter"
      onClick={[Function]}
      value="p"
    >
      p
    </div>
    <div
      className="letterBox toUseLetter"
      onClick={[Function]}
      value="l"
    >
      l
    </div>
    <div
      className="letterBox toUseLetter"
      onClick={[Function]}
      value="s"
    >
      s
    </div>
    <div
      className="letterBox toUseLetter"
      onClick={[Function]}
      value="b"
    >
      b
    </div>
    <div
      className="letterBox toUseLetter"
      onClick={[Function]}
      value="n"
    >
      n
    </div>
    <div
      className="letterBox toUseLetter"
      onClick={[Function]}
      value="o"
    >
      o
    </div>
    <div
      className="letterBox toUseLetter"
      onClick={[Function]}
      value="s"
    >
      s
    </div>
    <div
      className="letterBox toUseLetter"
      onClick={[Function]}
      value="i"
    >
      i
    </div>
  </div>
  <div
    className="hidden row"
  >
    matches:  
    0
  </div>
  <div
    className="row"
  />
</div>
`;
